### PR TITLE
[RISCV] Adjust select shuffle cost to reflect mask creation cost

### PR DESCRIPTION
**Pull request by preames**
This is inspired by https://github.com/llvm/llvm-project/pull/77342#pullrequestreview-1814673242, and is split off of same with some differences in style.

A select is a vmerge.vv with the additional cost of materializing the bitmask vector in a vreg.  All masks fit within a single vector register (e8 + m8 is the worst case), and thus our worst case cost should be roughly 3 (2 scalar to produce the address, one vector load op).  Given most shuffles are small, and the mask will be instead produced by LUI/ADDI + vmv.s.x or ADDI + vmv.s.x, using 2 as the default seems quite reasonable.  At worst, we're not going to be off by much.

The prior lowering scaled the cost of the bitmask with LMUL, which I don't understand.  At m1 it did use the same base cost of 2.  (@lukel97 You wrote the original code here, anything I'm missing here?)

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -443,10 +441,13 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
     // vmerge.vvm   v8, v9, v8, v0
+    // We use 2 for the cost of the mask materialization as this is the true
+    // cost for small masks and most shuffles are small.  At worst, this cost
+    // should be a very small constant for the constant pool load.  As such,
+    // we may bias towards large selects slightly more than truely warranted.
     return LT.first *
-           (TLI->getLMULCost(LT.second) + // FIXME: should be 1 for li
-            getRISCVInstructionCost({RISCV::VMV_S_X, RISCV::VMERGE_VVM},
-                                    LT.second, CostKind));
+           (2 + getRISCVInstructionCost({RISCV::VMERGE_VVM},
```
Comment (arcbbb) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:449]: Can we query the cost of VMV_S_*/VMV_*_S instead? So vendors can provide their own number in getRISCVInstructionCost.

**Comment by lukel97** [comment]
> The prior lowering scaled the cost of the bitmask with LMUL, which I don't understand. At m1 it did use the same base cost of 2. (@lukel97 You wrote the original code here, anything I'm missing here?)

Nope, looks like an oversight. AFAIK the vmerge.vvm is the only op that should be scaled by LMUL

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 1 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:449]: That makes sense. We would also need to address this FIXME in getRISCVInstructionCost too:
https://github.com/llvm/llvm-project/blob/e040252b545229041d0e51d397acd333936c5b42/llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp#L87-L90

Probably needs to be done in a separate patch.

**Comment by preames** [comment]
ping

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 1 above)
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:449]: @lukel97 Agree. Is is not blocking issue.
@preames I would like to override VMV_X_S for sifive-x280, and VMV_S_X is for consistency.

**Review by arcbbb — APPROVED** [review]
LGTM as change on VMV_S_\* can be a separate PR.

**Pull request merged**